worker/src/enclave/api.rs

```diff
@@ -183,214 +110,51 @@ pub fn enclave_init() -> SgxResult<SgxEnclave> {
         }
     }
 
-    let mut status = sgx_status_t::SGX_SUCCESS;
-    // call the enclave's init fn
-    let result = unsafe { init(enclave.geteid(), &mut status) };
-    if status != sgx_status_t::SGX_SUCCESS {
-        return Err(status);
-    }
-    if result != sgx_status_t::SGX_SUCCESS {
-        return Err(result);
-    }
-    Ok(enclave)
+    // create an enclave API and initialize it
+    let enclave_api = Enclave::new(enclave);
+    enclave_api.init()?;
+
+    Ok(enclave_api)
 }
 
-pub fn enclave_init_chain_relay(
-    eid: sgx_enclave_id_t,
+pub fn enclave_init_chain_relay<E: EnclaveApi>(
     genesis_header: Header,
     authority_list: VersionedAuthorityList,
     authority_proof: Vec<Vec<u8>>,
-) -> SgxResult<Header> {
-    let mut latest_header = vec![0u8; 200];
-
-    let mut status = sgx_status_t::SGX_SUCCESS;
-    let result = unsafe {
-        // Todo: this is a bit ugly but the common `encode()` is not implemented for authority list
-        authority_list.using_encoded(|authorities| {
-            init_chain_relay(
-                eid,
-                &mut status,
-                genesis_header.encode().as_ptr(),
-                genesis_header.encode().len(),
-                authorities.as_ptr(),
-                authorities.len(),
-                authority_proof.encode().as_ptr(),
-                authority_proof.encode().len(),
-                latest_header.as_mut_ptr(),
-                latest_header.len(),
-            )
-        })
-    };
+    enclave: &E,
+) -> EnclaveResult<Header> {
+    // Todo: this is a bit ugly but the common `encode()` is not implemented for authority list
+    let latest_header_encoded = authority_list.using_encoded(|authorities| {
+        enclave.init_chain_relay(
+            genesis_header.encode(),
+            authorities.to_vec(),
+            authority_proof.encode(),
+        )
+    })?;
 
-    if status != sgx_status_t::SGX_SUCCESS {
-        return Err(status);
-    }
-    if result != sgx_status_t::SGX_SUCCESS {
-        return Err(result);
-    }
-    let latest: Header = Decode::decode(&mut latest_header.as_slice()).unwrap();
+    let latest: Header = Decode::decode(&mut latest_header_encoded.as_slice()).unwrap();
```

Comment:
I can move the code. It would however mean, that the `enclave-api` crate will have dependencies on `substrate-node-runtime`. I'm not sure if that is problematic or not?